### PR TITLE
Remove watch commands, allow arbitrary expressions in Program State

### PR DESCRIPTION
**Pull request by Melamoto**
Does not include the actual removal of DexWatch (although it no longer does anything) or of DexWatch from tests, as these changes are trivial but amount to a very large diff which is unnecessary for a practical review of the change.

### CODE REVIEW

File: dex/debugger/visualstudio/VisualStudio.py [hunk 1]
```diff
@@ -155,19 +166,24 @@ def get_step_info(self):
...
 
-        loc = LocIR(**self._location)
-        if frames:
-            frames[0].loc = loc
-            state_frames[0].location = SourceLocation(**self._location)
+            if idx == 0:
+                frame.loc = loc
+                state_frame.location = SourceLocation(**self._location)
+
+            self.set_current_stack_frame(idx)
```
Comment (OCHyams) [inline dex/debugger/visualstudio/VisualStudio.py]: I may be missing something here. Could you remove this line (178) and call evaluate_expression with frame_idx=idx in the loop below?
```
for watch in self.watches:
    state_frame.watches[watch] = self.evaluate_expression(watch, idx)
```

**Comment by OCHyams** [comment]
How does this change interact with `annotate-expected-values`?
`annotate-expected-values` is used to generate `DexExpectWatchValues` and `DexExpectStepKind` commands based on evaluating `DexWatch` commands at -O0.

We might want to keep `DexWatch` commands for this purpose?

**Comment by Melamoto** [comment]
> 
> 
> Changes look good.
> 
> How does this change interact with `annotate-expected-values`?
> `annotate-expected-values` is used to generate `DexExpectWatchValues` and `DexExpectStepKind` commands based on evaluating `DexWatch` commands at -O0.
> 
> We might want to keep `DexWatch` commands for this purpose?

Good catch; I think `annotate-expected-values` will need to be changed to add line numbers to the DexExpectWatchValues command, as otherwise the value for the watched expression will be checked at every step of the program, including those where it is out of scope or uninitialized. I think the simple solution is to have it use the earliest line on which an expression is `DexWatch`ed as the from_line, and the last as the to_line. Thoughts?

**Comment by OCHyams** [comment]
> I think the simple solution is to have it use the earliest line on which an expression is DexWatched as the from_line, and the last as the to_line. Thoughts?

This, or maybe we should change `DexWatch` so that it doesn't have to be on each source line. I think it makes sense that `DexWatch` also takes the `from_line`, `to_line`, and `on_line` args.

I think that would make `DexWatch` easier to use?

If we did go down this path I reckon it might have to come as a separate patch. What do you think about this?

**Comment by Melamoto** [comment]
I think there's maybe a little bit too much speculation about what would be a good use for DexWatch, without much empirical evidence either way. I'd say it should be a separate patch, but that work should itself be contingent on a demonstrated need for it (especially considering that we're aiming to put this upstream soon).

**Comment by OCHyams** [comment]
Seems fair. In which case I think your original suggestion would be fine.

> I think the simple solution is to have it use the earliest line on which an expression is DexWatched as the from_line, and the last as the to_line.

@TomWeaver18, @jmorse, thoughts?

**Comment by TomWeaver18** [comment]
apart from small nit, LGTM

File: dex/debugger/DebuggerBase.py [hunk 2]
```diff
@@ -136,6 +137,10 @@ def start(self):
         self.steps.clear_steps()
         self.launch()
 
+        for command in chain.from_iterable(self.steps.commands.values()):
+            command_obj = get_command_object(command)
```
[the remark this answers is not in the document]
Comment (Melamoto) [reply dex/debugger/DebuggerBase.py]: That's something to resolve in either one of the two patches when the other goes in I think; all that's important is that at this line we have access to the command object.

**Comment by jmorse** [comment]
LGTM with some minor quibbles asked inline. This is going to result in all expressions used being evaluated in each stack frame, yes? (I think we've been over the impact of that so it should be fine).

**Comment by Melamoto** [comment]
> 
> 
> LGTM with some minor quibbles asked inline. This is going to result in all expressions used being evaluated in each stack frame, yes? (I think we've been over the impact of that so it should be fine).

Yes, that's the method we've gone with. Also, are you referring to quibbles by previous reviewers, or quibbles of your own making? (Asking because I can't see any inline comments added from you)

File: dex/debugger/lldb/LLDB.py [hunk 3]
```diff
@@ -177,9 +177,11 @@ def get_step_info(self):
             state_frame = StackFrame(function=frame.function,
                                      is_inlined=frame.is_inlined,
                                      location=SourceLocation(**loc_dict),
-                                     local_vars={})
-            for expr in sb_frame.GetVariables(True, True, True, True):
-                state_frame.local_vars[expr.name] = expr.value
+                                     watches={})
+            for expr in map(
+                lambda watch, idx=i: self.evaluate_expression(watch, idx),
```
[the remark this answers is not in the document]
Comment (Melamoto) [reply dex/debugger/lldb/LLDB.py]: Pylint complains about it - using `idx=i` binds `i`'s value to `idx` at the point that the lambda is created, whereas using `i` directly evaluates it at the time the lambda is run (which could cause unexpected and difficult to diagnose behaviour). While in this case we're executing the lambda immediately after creating it, this approach is just good practice as I see it.

File: dex/debugger/DebuggerBase.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (TomWeaver18) [reply dex/debugger/DebuggerBase.py]: sure, I meant this as an inline comment/headsup. not a review comment, my apologies

**Pull request merged**